bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/AdaptiveImageServlet.java

```diff
@@ -437,19 +440,77 @@ private void resizeAndStreamLayer(SlingHttpServletResponse response, Layer layer
     }
 
     /**
-     * Given an {@link Asset}, this method will return the first web {@link Rendition} it finds in the asset's renditions list.
+     * Given an {@link Asset}, this method will return the WCM rendition (cq5dam.web.*)
      *
      * @param asset the asset for which to retrieve the web rendition
-     * @return the rendition, if found, {@code null} otherwise
+     * @return the WCM rendition, if found the original
+     */
+    @NotNull
+    private EnhancedRendition getWCMRendition(@NotNull Asset asset) {
+        return new EnhancedRendition(asset.getRendition(new WCMRenditionPicker()));
+    }
+
+    /**
+     * Given an {@link Asset} and a specified width, this method will return the best rendition
+     * for that width (smallest rendition larger than the specified width) or the original.
+     *
+     * @param asset the asset for which to retrieve the best rendition
+     * @param width the width
+     * @return a rendition that is suitable for that width
+     * @throws IOException when the best suited rendition is too large for processing
      */
-    private Rendition getAWebRendition(Asset asset) {
-        List<Rendition> renditions = asset.getRenditions();
+    @NotNull
+    private EnhancedRendition getBestRendition(@NotNull Asset asset, int width) throws IOException {
+        // Sort renditions by file size
+        SortedSet<Rendition> renditions = new TreeSet<>((o1, o2) -> Long.valueOf(o1.getSize() - o2.getSize()).intValue());
+        renditions.addAll(asset.getRenditions());
+        EnhancedRendition bestRendition = null;
+        // Find first rendition that has a width larger or equal than wanted
         for (Rendition rendition : renditions) {
-            if (rendition.getName().startsWith(DamConstants.PREFIX_ASSET_WEB)) {
-                return rendition;
+            EnhancedRendition enhancedRendition = new EnhancedRendition(rendition);
+            Dimension dimension = enhancedRendition.getDimension();
+            if (dimension != null) {
+                if (dimension.getWidth() >= width) {
```

Comment:
Instead of throwing an `IOException` later in the filter call this method should already check if the requested with is already larger than the the maxInputWidth and should return the smales rendition which is larger than the requested width or maxInputWidth.
I would rather change this line to `dimension.getWidth() >= Math.min(width, maxInputWidth)

Reply:
The filter is used for both _bestRendition_ and _original_ (when no transformation is needed). As it is now, `getBestRendtion` will always return something (falling back to _original_). If we do like you proposed and there's no rendition between `width` and `maxInputWidth` it would have to return `null` and we would need to add checks throughout the rest of the logic, which would make the code more complex and harder to read, IMO.

Also, I think throwing a `IOException` is good and meaningful as it will signal there's an error with reading that image/rendition (it's too large).

Reply:
Okay, I got it. LGTM!